KLEEBuilder: ugly dirty band-aid hack to workaround recursion. Fixes #400

LebedevRI opened the pull request:
`KLEEBuilder::get()` is recursive.
It already has problems with running out of stack space
with very trivial inputs, since there are a lot of IR instructions
souper knows how to produce, and thus a lot of possible replacements.
(that were returned from `GetCandidateExprForReplacement()`)

But `KLEEBuilder::get()` has built-in caching.
And recursion only kicks in if the `Inst` is not found in cache.
Thus we simply need to *try* to pre-populate the cache.

The caveat here is that if we try to pre-generate `Inst` I,
it will need it's operands to be already generated,
and we are back into recursion business.

Thus, we need to collect all of the `Inst` Operands from every `Inst`
that is reachable from the current 'root' `Inst`, without recursion,
and do pre-generation in reverse order, first processing Defs,
and then Uses. so that when we get to `Inst` I that has Op0 and Op1 ops,
those ops must already be pre-generated.

Note that we can't really split `get()` into `getSimple()` and
`getOrBuildRecursive()` because we won't reach every single one of
these `Inst` because we e.g. won't precompute the `addnswUB()` & friends.

So while this certainly does not replace recursion with a proper work-list,
it should be sufficient to get rid of recursion in 99.99% of cases.

Another startling discovery i have made:
`ExprSMTLIBPrinter` produces SMT query in a string form?!
I have done some rudimentary benchmarks, and that seems to be giant
performance blackhole.
I'm really looking forward moving from KLEE to alive i guess :)

Revert "increase stack a different way"

This reverts commit c490eff970bcf9f447cf2e344a139024719e8f18.

Revert "increase stack for CI"

This reverts commit 89e3fdf445b9685f283541496a14311c7328ba3a.

Revert "increase stack for CI"

This reverts commit 78c8196629d1220a0f185ae1e36d90f5c169535b.

regehr commented:
Thanks Roman! I'll try to review soon. Yes, communicating with the solver using textual smtlib is really slow and we need to migrate to alive2 as soon as possible!
It's @manasij7479 who is working on this and perhaps I have not properly communicated to him the importance of making this happen.

LebedevRI commented:
> we need to migrate to alive2 as soon as possible!
> It's @manasij7479 who is working on this and perhaps I have not properly communicated to him the importance of making this happen.

If there is any sort of reasonable TO-DO list, do tell, i'd love to try to help.

regehr commented:
@manasij7479 is close to getting phi working-- let's revisit this soon, once that happens.

LebedevRI commented:
> @manasij7479 is close to getting phi working-- let's revisit this soon, once that happens.

("this" meant alive-in-souper (which i 100% understand.), or this pr (not particularly intrusive fix)?)

regehr commented:
ah, I plan to merge this PR soon-- I just want to fully understand it, this is a part of Souper that I've always left to others and I don't really understand it yet (ugh)

LebedevRI commented:
I've left hopefully-good comments, but let's do this once more:
1. We have the following tree: `sub(add(x, y), mul(z, 2))`
2. We want to get `ref<Expr>` for the topmost node, the `sub`. Thus we call `get()` on it.
3. We do not yet have an `Expr` for this `sub` in `ExprMap` cache, therefore we need to generate..
4. `sub` has two operands: `add` and `mul`. In order to produce the `Expr` for `sub`, we need to have `Expr` for `add` and `Expr` for `mul`.
5. Therefore we recurse back into step 2., with topmost node being `add`
6. We want to get `ref<Expr>` for the topmost node, the `add`. Thus we call `get()` on it.
7. We do not yet have an `Expr` for this `add` in `ExprMap` cache, therefore we need to generate..
8. `add` has two operands: `x` and `y`. They are simple variables, so no generation will be needed.
9. Now that we `Expr`s for `x` and `y`, we produce `Expr` for `add`.
10. Great, we now have `Expr` for `add`, but we don't have `Expr` for `mul`.
11. Therefore we recurse back into step 2., with topmost node being `mul`
12. We want to get `ref<Expr>` for the topmost node, the `mul`. Thus we call `get()` on it.
13. We do not yet have an `Expr` for this `mul` in `ExprMap` cache, therefore we need to generate..
14. `add` has two operands: `z` and `2`. They are simple variables, so no generation will be needed.
15. Now that we `Expr`s for `z` and `2`, we produce `Expr` for `mul`.
16. Now that we `Expr`s for `add` and `mul`, we produce `Expr` for `sub`.

Note how recursive this is: we call `get()` for `add` and for `mul`; from `get()` for `sub`.
This is the root problem that we need to avoid.

As it is evident, the recursion *only* happens when we are unable to find `Expr` in `ExprMap` cache.
Therefore, if we somehow manage to materialize all of the `Expr`s in `ExprMap` cache *before* we start looking for them in that cache, we will successfully avoid recursion.

The simple observation is that it's a tree. Given a node, we know it's operands:
* `sub(add(x, y), mul(z, 2))` has operands `add(x, y)` and `mul(z, 2)`
* `add(x, y)` has operands `x` and `y`
* `mul(x, y)` has operands `z` and `2`

Therefore, we need to somehow come up with a list *without recursion*.
Let's start at the root, add it to vector (next: worklist)
* `sub(add(x, y), mul(z, 2))`

And now, let's have a loop over that vector, but not over the elements of a vector, but over the position and size of the vector, since we will be appending(!) to the vector while iterating, and we do want to visit those appended elts too.
And now, for every element we are iterating over: from `build()` we can see that we will most likely
call `get()` for every element in `orderedOps()` next, in order. So just add every element from `orderedOps()` to our worklist we are iterating over. (FIXME: i'm not sure if we want to append `orderedOps()` elts, or `reverse(orderedOps())` elts).

Now that the loop over worklist has finished, the worklist contains the every node reachable from the
top-level node, in correct order:
* `sub(add(x, y), mul(z, 2))`
* `add(x, y)`
* `mul(z, 2)`
* `x`
* `y`
* `z`
* `2`

And now, let's loop over the final worklist, but in reverse order, end-to-beginning:
* `2` - a constant, trivially generated, added to cache
* `z` - a parameter, no generation needed.
* `y` - a parameter, no generation needed.
* `x` - a parameter, no generation needed.
* `mul(z, 2)` - both `2` and `z` were retrieved from cache, no recursion!, `mul` generated, added to cache.
* `add(x, y)` - both `x` and `z` were retrieved from cache, no recursion!, `add` generated, added to cache.
* `sub(add(x, y), mul(z, 2))` - both `add(x, y)` and `mul(z, 2)` were retrieved from cache, no recursion!, `sub` generated, added to cache.

Therefore, once we call `get()` on top-level `sub(add(x, y), mul(z, 2))` for real,
we will retrieve it from cache, with no recursion.
You could call this prefetching, cache heat-up, recursion avoidance, etc. 

Hopefully this makes sense:)
Caveat: i suspect, the prefetching will be blind on things like `get(addnuwUB(I))`. I'm not sure what will happen if there is a large chain of those.

regehr commented:
thanks Roman! this is great! just the bandaid we need :)
